File: audit/src/main/java/com/tek/audit/model/WebAudit.java
package com.tek.audit.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.tek.audit.model.json.AuditWebRequest;
import com.vladmihalcea.hibernate.type.json.JsonBinaryType;
import lombok.*;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.TypeDef;
import org.hibernate.annotations.TypeDefs;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import javax.persistence.*;
import java.time.Instant;
import java.util.Objects;

@NoArgsConstructor
@AllArgsConstructor
@Builder
@Data
@ToString(exclude = {"request"})

@TypeDefs({
    @TypeDef(name = "jsonb", typeClass = JsonBinaryType.class)
})

@Entity
@Table(name = "web_audit", schema = "audit")
@EntityListeners(AuditingEntityListener.class)
@JsonIgnoreProperties(value = {"createdAt", "updatedAt"}, allowGetters = true)
public class WebAudit {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "id")
  private Long id;

  @Type(type = "jsonb")
  @Column(name = "request", columnDefinition = "jsonb")
  @Basic
  private AuditWebRequest request;

  @Column(name = "status")
  private Integer status;

  @Column(name = "duration")
  private Long duration;

  @Column(name = "created_at")
  @CreatedDate
  private Instant createdAt;

  @Column(name = "updated_at")
  @LastModifiedDate
  private Instant updatedAt;

  @Override
  public boolean equals(Object o) {
    if (this == o) {return true;}
    if (o == null || getClass() != o.getClass()) {return false;}
    WebAudit webAudit = (WebAudit) o;
    return id.equals(webAudit.id);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id);
  }
}

File: audit/src/main/kotlin/com/tek/audit/model/WebAudit.kt
package com.tek.audit.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.javers.core.metamodel.annotation.TypeName
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.Instant
import java.util.*
import javax.persistence.*

@Entity
@TypeName("web_audit")
@EntityListeners(AuditingEntityListener::class)
@JsonIgnoreProperties(
    "requested_at", "served_at",
    allowGetters = true
)
@Table(name = "web_audit")
class WebAudit(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(columnDefinition = "text")
    val request: String,

    @Column(columnDefinition = "text")
    var response: String? = null,

    var stats: String? = null,

    @Transient
    val initTime: Long,

    @Transient
    val initTimeMillis: Long
) {

    @CreatedDate
    @Column(name = "requested_at")
    var requestedAt: Instant? = null

    @LastModifiedDate
    @Column(name = "served_at")
    var servedAt: Instant? = null
}




